src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -745,63 +746,204 @@ void StorageKafka2::dropReplica()
     }
 }
 
-std::optional<StorageKafka2::TopicPartitionLocks>
-StorageKafka2::lockTopicPartitions(zkutil::ZooKeeper & keeper_to_use, const TopicPartitions & topic_partitions)
-{
-    std::vector<fs::path> topic_partition_paths;
-    topic_partition_paths.reserve(topic_partitions.size());
-    for (const auto & topic_partition : topic_partitions)
-        topic_partition_paths.emplace_back(getTopicPartitionPath(topic_partition));
 
-    Coordination::Requests ops;
+/// We go through all the replicas, count the number of live replicas, 
+/// and see which partitions are already blocked by other replicas
+std::set<KafkaConsumer2::TopicPartition> StorageKafka2::lookupReplicaState(zkutil::ZooKeeper & keeper_to_use)
+{
+    LOG_TRACE(log, "Starting to lookup replica's state");
+    Strings replicas;
+    if (Coordination::Error::ZOK != keeper_to_use.tryGetChildren(keeper_path + "/replicas", replicas))
+        return {};
 
-    for (const auto & topic_partition_path : topic_partition_paths)
+    ActiveReplicaCount = 0;
+    std::set<KafkaConsumer2::TopicPartition> locked_partitions;
+    for (const auto& replica : replicas)
     {
-        const auto lock_file_path = String(topic_partition_path / lock_file_name);
+        String current_replica_path = keeper_path + "/replicas/" + replica;
+        String node_data;
+        if (!keeper_to_use.tryGet(current_replica_path + "/topics_assigned", node_data))
+            continue;
+        LOG_INFO(log, "Current state of the replica {} is \n {}", replica, node_data);
+        ReplicaStatePtr replica_state = ReplicaState::parse(node_data);
+        if (current_replica_path == replica_path)
+        {
+            ReplicaState.permanent_topic_partitions = replica_state->permanent_topic_partitions;
+            ReplicaState.permanent_topics_assigned = replica_state->permanent_topics_assigned;
+            ReplicaState.tmp_topic_partitions = replica_state->tmp_topic_partitions;
+            ReplicaState.tmp_topics_assigned = replica_state->tmp_topics_assigned;
+            LOG_INFO(log, "Local state of the replica {} was updated successfully", replica);
+        }
+        ++ActiveReplicaCount;
+        locked_partitions.insert(replica_state->permanent_topics_assigned.begin(), replica_state->permanent_topics_assigned.end());
+        locked_partitions.insert(replica_state->tmp_topics_assigned.begin(), replica_state->tmp_topics_assigned.end());
+    }
+
+    return locked_partitions;
+}
+
 
-        // It is okay that these paths are created in a different transaction. The important thing is the lock file.
+void StorageKafka2::createLocksInfo(zkutil::ZooKeeper & keeper_to_use, TopicPartitionLocks & locks, const TopicPartition& partition_to_lock, bool new_lock)
+{
+    const auto topic_partition_path = getTopicPartitionPath(partition_to_lock);
+    if (new_lock) {
```

Comment:
This logic is not correct. You cannot just create an emphemeral node holder without actually having an emphemeral node. It is somewhat similar to a unique pointer. I will try to explain this more in private conversation to make sure we are on the same page.